js/data/feature_tree.js

```diff
@@ -88,80 +152,181 @@ FeatureTree.prototype.query = function(args, styleLayersByID) {
         maxY = Math.max(maxY, p.y);
     }
 
-    var bounds = [minX - additionalRadius, minY - additionalRadius, maxX + additionalRadius, maxY + additionalRadius];
-    var symbolQueryBox = new CollisionBox(new Point(minX, minY), 0, 0, maxX - minX, maxY - minY, args.scale, null);
+    var matching = this.grid.query(minX - additionalRadius, minY - additionalRadius, maxX + additionalRadius, maxY + additionalRadius);
+    matching.sort(topDown);
+    var match = this.featureIndexArray.get(0);
+    this.filterMatching(result, matching, match, queryGeometry, filter, params.layers, styleLayersByID, args.bearing, pixelsToTileUnits, returnGeoJSON);
+
+    var matchingSymbols = this.collisionTile.queryRenderedSymbols(minX, minY, maxX, maxY, args.scale);
+    var match2 = this.collisionTile.collisionBoxArray.get(0);
+    matchingSymbols.sort();
+    this.filterMatching(result, matchingSymbols, match2, queryGeometry, filter, params.layers, styleLayersByID, args.bearing, pixelsToTileUnits, returnGeoJSON);
+
+    return result;
+};
+
+function topDown(a, b) {
```

Comment:
I prefer `topDown` because the intention is to sort the keys by the order their features are drawn, which just happens to be reverse/descending order. There's probably an even clearer name I'm not thinking of right now.